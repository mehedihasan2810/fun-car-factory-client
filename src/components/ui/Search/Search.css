.search-form-control {
    display: flex;
    align-items: center;
    border-radius: 100vw;
    overflow: hidden;
    background-color: var(--primary-sky-800);
    padding-inline-start: 1rem;
  }
  
  .search-form-control svg {
    display: inline-block;
    width: 25px;
    height: 25px;
    margin-inline-end: 4px;
  }
  
  .search-form-control svg,
  .search-form-control input::placeholder {
    color: var(--base-gray-300);
  }
  
  .search-form-control:focus-within {
    outline: 1px solid var(--primary-sky-500);
  }
  
  .search-form-control :is(input, button) {
    padding-block: 1rem;
    border: none;
    background-color: transparent;
  }
  .search-form-control :is(input, button):focus {
    outline: 0;
  }
  .search-form-control input {
    inline-size: 400px;
    color: var(--base-gray-50);
  }
  .search-form-control button {
    cursor: pointer;
    background-color: var(--base-gray-200);
    background-color: var(--primary-sky-700);
    padding-inline: 1rem;
    color: var(--primary-sky-100);
  }
  .search-form-control button:hover{
    background-color: var(--primary-sky-600);
  }
  