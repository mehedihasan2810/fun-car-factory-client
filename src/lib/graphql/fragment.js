import gql from "graphql-tag";

export const CORE_CARS_FIELD = gql`
  fragment CoreCarsField on Car {
    id
    name
    price
    url
    rating
    category
    quantity
  }
`;
