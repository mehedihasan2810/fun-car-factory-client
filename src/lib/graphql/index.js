import {
  ApolloClient,
  InMemoryCache,
  createHttpLink,
  from,
} from "@apollo/client";
import { setContext } from "@apollo/client/link/context";
import { onError } from "@apollo/client/link/error";
import { signOut } from "firebase/auth";
import Cookies from "js-cookie";
import { auth } from "../../configs/firebase";

// Create an in-memory cache
export const cache = new InMemoryCache();

// Create an HTTP link to the GraphQL server
const httpLink = createHttpLink({
  uri: "https://fun-car-factory-server.vercel.app/graphql",
  // uri: "http://localhost:4000/graphql",
});

// Set the authorization context based on the user's token
const authLink = setContext((_, { headers }) => {
  const token = Cookies.get("token");

  return {
    headers: {
      ...headers,
      authorization: token ? `Bearer ${token}` : "",
    },
  };
});

// Handle errors, including authentication errors
const errorLink = onError(async ({ graphQLErrors, networkError }) => {
  if (graphQLErrors) {
    if (graphQLErrors[0]?.message === "User is not authenticated") {
      // Sign out the user, remove the token, and clear the Apollo Client store
      await signOut(auth);

      Cookies.remove("token");

      apolloClient.clearStore();
    }

    // Log GraphQL errors
    graphQLErrors.forEach(({ message }) =>
      console.error(`[GraphQL error]: Message: ${message}`)
    );
  }
  if (networkError) console.error(`[Network error]: ${networkError}`);
});

// Create the Apollo Client with the configured link and cache
export const apolloClient = new ApolloClient({
  // link: authLink.concat(errorLink, httpLink),
  link: from([authLink, errorLink, httpLink]),
  cache,
});
