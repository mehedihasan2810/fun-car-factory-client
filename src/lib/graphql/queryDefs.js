import { gql } from "@apollo/client";

export const GET_CARS = gql`
  query GetCars {
    getCars {
      id
      name
      price
      url
      rating
    }
  }
`;

export const GET_CAR = gql`
  query GetCar($id: String!) {
    getCar(id: $id) {
      id
      name
      price
      rating
      url
      sellerName
      email
      category
      description
    }
  }
`;
