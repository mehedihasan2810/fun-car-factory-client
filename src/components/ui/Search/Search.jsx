import { useState } from "react";
import "./Search.css";

const Search = ({ onHandleSearch, placeholder }) => {
  const [search, setSearch] = useState("");

  return (
    <div className="search-form-control">
      <svg
        xmlns="http://www.w3.org/2000/svg"
        fill="none"
        viewBox="0 0 24 24"
        strokeWidth={1.5}
        stroke="currentColor"
        className="w-6 h-6"
      >
        <path
          strokeLinecap="round"
          strokeLinejoin="round"
          d="M21 21l-5.197-5.197m0 0A7.5 7.5 0 105.196 5.196a7.5 7.5 0 0010.607 10.607z"
        />
      </svg>

      <input
        value={search}
        onChange={(e) => setSearch(e.target.value)}
        onKeyDown={(e) => {
          if (e.key === "Enter") {
            onHandleSearch(search);
          }
        }}
        type="text"
        id="search"
        name="search"
        placeholder={placeholder}
      />
      <button onClick={() => onHandleSearch(search)} type="submit">
        Search
      </button>
    </div>
  );
};

export default Search;
