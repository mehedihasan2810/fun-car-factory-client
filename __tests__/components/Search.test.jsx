import { ApolloProvider } from "@apollo/client";
import {
  customRender,
  describe,
  expect,
  screen,
  test,
  userEvent,
} from "../utils/utils";
import Search from "../../src/components/ui/Search/Search";
import { apolloClient } from "../../src/lib/graphql";

describe("<Search />", () => {
  test("Should render successfully and match snapshot", () => {
    const { asFragment } = customRender(
      <ApolloProvider client={apolloClient}>
        <Search />
      </ApolloProvider>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test("Search input field should work as espected", async () => {
    const user = userEvent.setup();

    customRender(<Search />);

    const searchInput = screen.getByTestId("search-input");

    expect(searchInput).toBeInTheDocument();

    await user.type(searchInput, "ferrari toys");

    expect(searchInput).toHaveValue("ferrari toys");
  });
});
