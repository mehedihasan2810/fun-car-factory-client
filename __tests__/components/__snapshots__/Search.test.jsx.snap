// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Search /> > Should render successfully and match snapshot 1`] = `
<DocumentFragment>
  <div
    class="search-form-control"
  >
    <svg
      class="w-6 h-6"
      fill="none"
      stroke="currentColor"
      stroke-width="1.5"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M21 21l-5.197-5.197m0 0A7.5 7.5 0 105.196 5.196a7.5 7.5 0 0010.607 10.607z"
        stroke-linecap="round"
        stroke-linejoin="round"
      />
    </svg>
    <input
      data-testid="search-input"
      id="search"
      name="search"
      type="text"
      value=""
    />
    <button
      data-testid="search-btn"
      type="submit"
    >
      Search
    </button>
  </div>
</DocumentFragment>
`;
